### ISSUE TITLE
Support for installing Mono via winetricks?

### ISSUE THREAD

**Issue by kenorb** [Winetricks/winetricks#1236]
Can Winetricks install [Mono MSI](http://dl.winehq.org/wine/wine-mono/4.8.3/) instead of installing it manually as per [Wiki](https://wiki.winehq.org/Mono)?

I'm aware it can `remove_mono`, but I can't see option to install it. I can see the Mono popup asking to install it when running `wineboot -u`, or by overriding `WINEDLLOVERRIDES`, e.g. `WINEDLLOVERRIDES=fixme+all winetricks -v`, but I'm looking for non-interactive way with minimum hassle (like `winetricks mono`).

Is there any reason why Mono is not on the list of libraries to install? I think in the past Mono was supported. Was there some reason to remove it?

**Comment by DarkShadow44** [comment]
Wine automatically installs Mono, the Wiki even lists the search paths.

**Comment by kenorb** [comment]
It doesn't install automatically, it shows the popup with the installer.
![image](https://user-images.githubusercontent.com/266306/58817311-618b3080-8623-11e9-81c4-907fbe14d4dd.png)
Which is not what I'm looking for.

**Comment by DarkShadow44** [comment]
Because, as it says, it doesn't find a Mono package. Put the right msi installer in the right folder, and it'll install automatically.

For the path, see the wiki.
For the mono version, see dlls/appwiz.cpl/addons.c of the version of wine you're using.

**Comment by austin987** [comment]
Fwiw: https://bugs.winehq.org/show_bug.cgi?id=47316

**Comment by a-berg** [comment]
> Because, as it says, it doesn't find a Mono package. Put the right msi installer in the right folder, and it'll install automatically.
> 
> For the path, see the wiki.
> For the mono version, see dlls/appwiz.cpl/addons.c of the version of wine you're using.

I think the OP was referring to that already, that he/she doesn't want to use that "manual" method of copying the msi to a folder, and instead would like to do something like `winetricks mono` and have it done automatically

**Comment by Botspot** [comment]
Any updates on this? I'm interested in doing the same thing.

**Comment by austin987** [comment]
No updates. The main problem is that I can't see an easy way to figure out what version of mono a given wine version needs without hard coding it (i.e., the value isn't in the registry/plain text anywhere, just embedded in mscoree.dll/appwiz.cpl).

And given that wine automatically gets mono on first boot/has an automated option already, it's a lower priority.

**Comment by Botspot** [comment]
> And given that wine automatically gets mono on first boot/has an automated option already, it's a lower priority.

On my system, when updating a prefix, wine displays two pop-up windows asking if I want to download Mono and asking if I want to download Gecko. This might not seem like a big deal for one wine prefix, but when managing many prefixes, it's very boring to click OK hundreds of times... There has got to be a way to install Mono and Gecko unattended.

**Comment by austin987** [comment]
There is:
https://wiki.winehq.org/Gecko

TL;DR: put the .msi files for gecko/mono in `~/.cache/wine`.

**Comment by Botspot** [comment]
> TL;DR: put the .msi files for gecko/mono in `~/.cache/wine`.

Would it be possible to place those files in a universal location, rather than in one user's home directory?
And yes, I've seen the wiki, and how it says to place files in `/usr/share/wine/gecko`, problem is that folder doesn't exist. I have manually installed wine to `/opt/wine`. (So the wine command is a symlink to `/opt/wine/bin/wine`)
In my case, where would I place the .msi file?

**Comment by austin987** [comment]
Look closer at the wiki:
`If you installed Wine in some $prefix rather than /usr, $prefix/share/wine/gecko/ before /usr/share/wine/gecko. (e.g. if you installed it from source, then place the files in /usr/local/share/wine/gecko).`

(or if you have wine in /usr, then make that directory and put the msi files there).

**Comment by Botspot** [comment]
> Look closer at the wiki: `If you installed Wine in some $prefix rather than /usr, $prefix/share/wine/gecko/ before /usr/share/wine/gecko. (e.g. if you installed it from source, then place the files in /usr/local/share/wine/gecko).`
> 
> (or if you have wine in /usr, then make that directory and put the msi files there).

After some more investigation, I have no idea what is happening!
After trial and error, I found that these commands will prevent the Mono and Gecko dialogs from appearing when booting a prefix for the first time. (Presumably that means that Gecko and Mono are installed automatically)
```
#mono
mkdir -p /opt/wine/mono
wget -O /opt/wine/mono.tar.xz "https://dl.winehq.org/wine/wine-mono/5.1.0/wine-mono-5.1.0-x86.tar.xz"
tar -xf /opt/wine/mono.tar.xz -C /opt/wine/mono
rm /opt/wine/mono.tar.xz

#gecko
mkdir -p /opt/wine/gecko
wget -O /opt/wine/gecko/gecko.tar.bz2 "https://dl.winehq.org/wine/wine-gecko/2.47.1/wine-gecko-2.47.1-x86.tar.bz2" 
tar -xf /opt/wine/gecko/gecko.tar.bz2 -C /opt/wine/gecko
rm /opt/wine/gecko/gecko.tar.bz2
```
Booting a Wine prefix now finishes completely unattended! :)
Except for one thing: I don't think gecko or mono are fully installed.
Here's a screenshot of `wine uninstaller`:
![2022-01-10-103707_1920x1080_scrot](https://user-images.githubusercontent.com/54716352/148862557-08d7865d-b0a0-4729-9870-cae05613f954.png)
As you can see, there is only one item in the list. Strange.

After deleting `/opt/wine/mono` and `/opt/wine/gecko`, I tried booting a new wine prefix. Two dialogs appeared, asking to install gecko and mono. I clicked Install for each one. After wineboot finished, here's what `wine uninstaller` looked like:
![2022-01-10-190852_1920x1080_scrot](https://user-images.githubusercontent.com/54716352/148863579-ec7d6078-9e81-4e0a-a4fa-16a249d8d7cd.png)
Any idea what is going on here?

**Comment by Botspot** [comment]
I figured it out. Turns out you had to download the .msi files to /opt/wine/gecko and /opt/wine/mono, *and retain the original filenames*!
Everything has been sorted out, no need to reply.

**Comment by austin987** [comment]
I don't think this is likely to be implemented, without changes upstream (see https://github.com/Winetricks/winetricks/issues/1236#issuecomment-1008542987)

**Comment by jflemer** [comment]
I understand this ticket is closed, but for future implementation in winetricks, or for others who end up here...

The script below works reasonably well on Ubuntu 22.04 with Wine 7.22 (`winehq-devel`). The detection of the wine install prefix dir might not work everywhere.
https://gist.github.com/jflemer/a1125763ea3fdb1fde2ca296675c7039

**Comment by austin987** [comment]
> I understand this ticket is closed, but for future implementation in winetricks, or for others who end up here...
> 
> The script below works reasonably well on Ubuntu 22.04 with Wine 7.22 (`winehq-devel`). The detection of the wine install prefix dir might not work everywhere. https://gist.github.com/jflemer/a1125763ea3fdb1fde2ca296675c7039

A neat idea, but unfortunately, not portable. `strings -e` seems to be a GNU extension. Running strings on macOS doesn't show the version string, afaict:
```
$ strings /Applications/Wine\ Devel.app/Contents/Resources/wine/lib/wine//i386-windows/appwiz.cpl | grep -i -e msi -e gecko -e mono
MsiInstallProduct failed: %lu
No MSI file
http://source.winehq.org/winegecko.php
GeckoUrl
GeckoCabDir
http://source.winehq.org/winemono.php
MonoUrl
MonoCabDir
FOLDERID_CommonOEMLinks
IID_ITfLangBarItemSink
__wine_delay_load_msi
__wine_import_msi_desc
__wine_import_msi_name
MsiInstallProductW
```

**Comment by jflemer** [comment]
Yeah, quite true, it is a GNU extension. So... less pretty (and slightly more error prone), but portable:
```
grep -a -o 'w.i.n.e.-.m.o.n.o.-.[^-]*-.[^.]*..m.s.i.' /opt/wine-devel/lib64/wine/x86_64-windows/appwiz.cpl | sed 's/\o000//g'
```
yields:
```
wine-mono-7.4.0-x86.msi
```

**Comment by austin987** [comment]
That grep doesn't work on BSD grep for me:

```
$ grep --version
grep (BSD grep, GNU compatible) 2.6.0-FreeBSD
$ grep -a 'w.i.n.e.-.m.o.n.o.-.[^-]*-.[^.]*..m.s.i.' appwiz.cpl ; echo $?
1
$ grep -a 'w.i.n.e.-.m.o.n.o.' appwiz.cpl  ; echo $?
1 
```

**Comment by WheezyE** [comment]
fwiw, I'm also interested in winetricks verbs for `winemono` and `winegecko`.

For example, winegecko is required for the winetricks `autohotkey` verb to install quietly. Other apps may require dotnetxx or winemono as dependencies and it would be great to check for those too.

No promises at all, but if I do ever learn more about these I'll try to report ideas here
